plugins/modules/ntnx_volume_groups.py

```diff
@@ -0,0 +1,413 @@
+            wait_for_task_completion(module, {"task_uuid": task_uuid})
+
+        vms = volume_group.read(volume_group_uuid, endpoint="/vm-attachments")
+        result["response"]["vms"] = vms.get("data")
+
+    # attach clients
+    if module.params.get("clients"):
+        for client in module.params["clients"]:
+
+            spec, err = Clients.get_spec(client, module.params.get("CHAP_auth"))
+            if err:
+                result["warning"] = "Client is not attached. Error: {0}".format(err)
+                result["skipped"] = True
+                continue
+
+            attach_resp = volume_group.attach_iscsi_client(spec, volume_group_uuid)
+
+            task_uuid = attach_resp["task_uuid"]
+            wait_for_task_completion(module, {"task_uuid": task_uuid})
+
+        clients = volume_group.read(
+            volume_group_uuid, endpoint="/iscsi-client-attachments"
+        )
+        result["response"]["clients"] = clients.get("data")
+
+
+# def update_volume_group(module, result):
+#     volume_group = VolumeGroup(module)
+#     volume_group_uuid = module.params.get("volume_group_uuid")
+#     if not volume_group_uuid:
+#         result["error"] = "Missing parameter volume_group_uuid in playbook"
+#         module.fail_json(msg="Failed updating volume_group", **result)
+#     result["volume_group_uuid"] = volume_group_uuid
+#
+#     # read the current state of volume_group
+#     resp = volume_group.read(volume_group_uuid)
+#     resp = resp.get("volume_group")
+#
+#     # new spec for updating volume_group
+#     update_spec, error = volume_group.get_spec(resp)
+#     if error:
+#         result["error"] = error
+#         module.fail_json(msg="Failed generating volume_group update spec", **result)
+#
+#     # check for idempotency
+#     if resp == update_spec:
+#         result["skipped"] = True
+#         module.exit_json(
+#             msg="Nothing to change. Refer docs to check for fields which can be updated"
+#         )
+#
+#     if module.check_mode:
+#         result["response"] = update_spec
+#         return
+#
+#     # update volume_group
+#     volume_group.update(update_spec, uuid=volume_group_uuid, no_response=True)
+#
+#     resp = volume_group.read(volume_group_uuid)
+#
+#     result["changed"] = True
+#     result["response"] = resp
+
+
+def delete_volume_group(module, result):
+    def detach_iscsi_clients():
+        clients_resp = volume_group.read(
+            volume_group_uuid, endpoint="/iscsi-client-attachments"
+        )
+        detached_clients = []
+        for client in clients_resp.get("data", []):
+
+            detach_resp = volume_group.detach_iscsi_client(volume_group_uuid, client)
+
+            task_uuid = detach_resp["data"]["extId"][-36:]
+            wait_for_task_completion(module, {"task_uuid": task_uuid})
+            detached_clients.append(client["extId"])
+
+        result["detached_clients"] = detached_clients
+
+    def detach_vms():
```

Comment:
We can detach multiple VMs and iscsi clients in delete ops. If every single vm or iscsi detachment is independent operation then we need to handle case where detachments of all entities should be attempted and should fail at last incase some ops on some entities fail in between. Suppose 1 VM detachment fails in between then we can error out after ops on all VMs have been done.